# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: browser/components/downloads/DownloadsMacFinderProgress.jsm
# Commit: 8287c3d0c106
# Full Hash: 8287c3d0c1064096ee7c162060e6615d72722ee0
# Author: Christoph Walcher <christoph-wa@gmx.de>
# Date: 2019-06-20 09:46:31
# Regressor Bug: 909760
# File Overlap Count: 1
# Description:
#   Bug 909760 - Show download progress in the MacOS Finder r=spohl,mak
#   
#   Show download progress in the MacOS Finder
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D30688
# ==============================================================================

diff -r 5b9a3de04646 -r 8287c3d0c106 browser/components/downloads/DownloadsMacFinderProgress.jsm
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/browser/components/downloads/DownloadsMacFinderProgress.jsm	Wed Jun 19 18:48:49 2019 +0000
@@ -0,0 +1,87 @@
+/* -*- indent-tabs-mode: nil; js-indent-level: 2 -*- */
+/* vim: set ts=2 et sw=2 tw=80 filetype=javascript: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Handles the download progress indicator of the macOS Finder.
+ */
+
+"use strict";
+
+var EXPORTED_SYMBOLS = [
+  "DownloadsMacFinderProgress",
+];
+
+const {XPCOMUtils} = ChromeUtils.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetters(this, {
+  Downloads: "resource://gre/modules/Downloads.jsm",
+});
+
+var DownloadsMacFinderProgress = {
+  /**
+   * Maps the path of the download, to the according progress indicator instance.
+   */
+  _finderProgresses: null,
+
+  /**
+   * This method is called after a new browser window on macOS is opened, it
+   * registers for receiving download events for the progressbar of the Finder.
+   */
+  register() {
+    // Ensure to register only once per process and not for every window.
+    if (!this._finderProgresses) {
+      this._finderProgresses = new Map();
+      Downloads.getList(Downloads.ALL).then(list => list.addView(this));
+    }
+  },
+
+  onDownloadAdded(download) {
+    if (download.stopped) {
+      return;
+    }
+
+    let finderProgress = Cc["@mozilla.org/widget/macfinderprogress;1"]
+      .createInstance(Ci.nsIMacFinderProgress);
+
+    let path = download.target.path;
+
+    finderProgress.init(path, () => {
+      download.cancel().catch(Cu.reportError);
+      download.removePartialData().catch(Cu.reportError);
+    });
+
+    if (download.hasProgress) {
+      finderProgress.updateProgress(download.currentBytes, download.totalBytes);
+    } else {
+      finderProgress.updateProgress(0, 0);
+    }
+    this._finderProgresses.set(path, finderProgress);
+  },
+
+  onDownloadChanged(download) {
+    let path = download.target.path;
+    let finderProgress = this._finderProgresses.get(path);
+    if (!finderProgress) {
+      // The download is not tracked, it may have been restarted,
+      // thus forward the call to onDownloadAdded to check if it should be tracked.
+      this.onDownloadAdded(download);
+    } else if (download.stopped) {
+      finderProgress.end();
+      this._finderProgresses.delete(path);
+    } else {
+      finderProgress.updateProgress(download.currentBytes, download.totalBytes);
+    }
+  },
+
+  onDownloadRemoved(download) {
+    let path = download.target.path;
+    let finderProgress = this._finderProgresses.get(path);
+    if (finderProgress) {
+      finderProgress.end();
+      this._finderProgresses.delete(path);
+    }
+  },
+};